cpp/src/arrow/dataset/scanner.cc

```diff
@@ -224,5 +225,100 @@ Result<std::shared_ptr<Table>> Scanner::ToTable() {
                                   FlattenRecordBatchVector(std::move(state->batches)));
 }
 
+struct ToBatchesState {
+  explicit ToBatchesState(size_t n_tasks)
+      : batches(n_tasks), task_drained(n_tasks, false) {}
+
+  /// Protecting mutating accesses to batches
+  std::mutex mutex;
+  std::vector<std::deque<std::shared_ptr<RecordBatch>>> batches;
+  std::vector<bool> task_drained;
+  size_t pop_cursor = 0;
+
+  void Push(std::shared_ptr<RecordBatch> b, size_t i_task) {
+    std::lock_guard<std::mutex> lock(mutex);
+    if (batches.size() <= i_task) {
+      batches.resize(i_task + 1);
+      task_drained.resize(i_task + 1);
+    }
+    batches[i_task].push_back(std::move(b));
+  }
+
+  Status Finish(size_t position) {
+    std::lock_guard<std::mutex> lock(mutex);
+    task_drained[position] = true;
+    return Status::OK();
+  }
+
+  std::shared_ptr<RecordBatch> Pop() {
+    std::unique_lock<std::mutex> lock(mutex);
+    std::condition_variable().wait_for(lock, std::chrono::milliseconds{1}, [this] {
```

Comment:
This seems a little off.  It looks like you're saying "If the current batch is still being filled then wait up to 1ms for it to add a new item."  But what happens if that 1ms expires?  Also, it looks like you're ignoring the return of `wait_for`.

For example, let's pretend that there are two scan tasks.

Scan task 1 RB 1 arrives at 10ms
Scan task 1 RB 1 arrives at 20ms
Scan task 2 RB 1 arrives at 30ms

The consumer grabs scan tasks very quickly.  So they come in at timestamp 15ms and they try to `Pop`.  batches[0].  It will sit in this loop for about 1ms (because tasks_drained[0] == false) and then break out.  Then pop_cursor == batches.size() will be true and so it will return null.

Reply:
This is indeed an (embarrassing) error. The intent was to throttle checking the variable to 1ms intervals but I forgot to write the enclosing loop. I'll replace with a simple call to `wait()`